File: Editorial/1001-1100/1055_Shortest_Way_to_Form_String/CSharp/2_Concatenate_until_Subsequence.cs
class Solution {
    public int ShortestWay(string source, string target) {

        // Boolean array to mark all characters of source
        bool[] sourceChars = new bool[26];
        foreach (char c in source) {
            sourceChars[c - 'a'] = true;
        }

        // Check if all characters of target are present in source
        // If any character is not present, return -1
        foreach (char c in target) {
            if (!sourceChars[c - 'a']) {
                return -1;
            }
        }

        // Concatenate source until target is a subsequence of concatenated string
        string concatenatedSource = source;
        int count = 1;
        while (!IsSubsequence(target, concatenatedSource)) {
            concatenatedSource += source;
            count++;
        }

        // Number of concatenations done
        return count;
    }

    // To check if toCheck is a subsequence of inString
    bool IsSubsequence(string toCheck, string inString) {
        int i = 0, j = 0;
        while (i < toCheck.Length && j < inString.Length) {
            if (toCheck[i] == inString[j]) {
                i++;
            }
            j++;
        }

        return i == toCheck.Length;
    }
}

File: Editorial/1001-1100/1055_Shortest_Way_to_Form_String/Java/2_Concatenate_until_Subsequence.java
class Solution {
    public int shortestWay(String source, String target) {

        // Boolean array to mark all characters of source
        boolean[] sourceChars = new boolean[26];
        for (char c : source.toCharArray()) {
            sourceChars[c - 'a'] = true;
        }

        // Check if all characters of the target are present in the source
        // If any character is not present, return -1
        for (char c : target.toCharArray()) {
            if (!sourceChars[c - 'a']) {
                return -1;
            }
        }

        // Concatenate source until the target is a subsequence of the concatenated string
        String concatenatedSource = source;
        int count = 1;
        while (!isSubsequence(target, concatenatedSource)) {
            concatenatedSource += source;
            count++;
        }

        // Number of concatenations done
        return count;
    }

    // To check if toCheck is a subsequence of the inString
    public boolean isSubsequence(String toCheck, String inString) {
        int i = 0, j = 0;
        while (i < toCheck.length() && j < inString.length()) {
            if (toCheck.charAt(i) == inString.charAt(j)) {
                i++;
            }
            j++;
        }

        return i == toCheck.length();
    }
}
